clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -796,13 +798,23 @@ final synchronized boolean hasValidMemberId() {
     }
 
 
+    public synchronized void resetGeneration(String causeMessage, boolean dueToError) {
+        if (dueToError) {
+            log.debug("Resetting generation and requesting re-join group due to {}", causeMessage);
+        } else {
+            log.debug("Resetting generation pro-actively due to {}", causeMessage);
+        }
+
+        this.generation = Generation.NO_GENERATION;
+        this.state = MemberState.UNJOINED;
+        this.rejoinNeeded = true;
+    }
+
     /**
      * Reset the generation and memberId because we have fallen out of the group.
      */
-    protected synchronized void resetGeneration() {
-        this.generation = Generation.NO_GENERATION;
-        this.rejoinNeeded = true;
-        this.state = MemberState.UNJOINED;
+    protected synchronized void resetGeneration(String errorMessage) {
```

Comment:
The thing is that resetGeneration is also called from `maybeLeaveGroup` for consumer's proactively resetting generation. This signature would not cover this case. Or are you suggesting we just have two independent call trace for these two cases?

And thinking about this, I think we should call resetGeneration before sending leave-group because we need to make sure `partitonRevoked` is called before the new rebalance is triggered. On the other hand, however, `maybeAutoCommitOffsetsSync` would have committed the offsets, so if user's `onPartitionsRevoked` callback is just for committing offsets, then it would be double committing if it is not checking `committed == position` in practice... the latter would only be remedied by more education I guess.